pulser-core/pulser/pulse.py

```diff
@@ -178,12 +185,64 @@ def ConstantPulse(
             detuning: The detuning value (in rad/µs).
             phase: The pulse phase (in radians).
             post_phase_shift: Optionally lets you add a
-                phase shift (in rads) immediately after the end of the pulse.
+                phase shift (in rad) immediately after the end of the pulse.
         """
         amplitude_wf = ConstantWaveform(duration, amplitude)
         detuning_wf = ConstantWaveform(duration, detuning)
         return cls(amplitude_wf, detuning_wf, phase, post_phase_shift)
 
+    @classmethod
+    @parametrize
+    def ArbitraryPhase(
+        cls,
+        amplitude: Waveform | Parametrized,
+        phase: Waveform | Parametrized,
+        post_phase_shift: float | Parametrized = 0.0,
+    ) -> Pulse:
+        r"""Pulse with an arbitrary phase waveform.
+
+        Due to how the Hamiltonian is defined in Pulser, the phase and
+        detuning are related by
+
+        .. math:: \phi(t) = \phi_c - \sum_{k=0}^{t} \delta(k)
+
+        where :math:`\phi_c` is the pulse's constant phase offset.
+        From a given phase waveform, we extract the phase offset and detuning
+        waveform that respect this formula for every sample of :math:`\phi(t)`
+        and use these quantities to define the Pulse.
+
+        Warning:
+            Expect when the phase waveform is a ``ConstantWaveform`` or a
```

Comment:
I think what you meant was rather except :) 
```suggestion
            Expect when the phase waveform is a ``ConstantWaveform`` or a
```
```suggestion
            Except when the phase waveform is a ``ConstantWaveform`` or a
```